mobile/src/main/java/org/openhab/habdroid/model/OpenHABWidget.java

```diff
@@ -10,293 +10,329 @@
+                .legend(widgetJson.has("legend") ? widgetJson.getBoolean("legend") : null)
+                .height(widgetJson.optInt("height"))
+                .iconColor(parseColor(widgetJson.optString("iconcolor", null)))
+                .labelColor(parseColor(widgetJson.optString("labelcolor", null)))
+                .valueColor(parseColor(widgetJson.optString("valuecolor", null)))
+                .encoding(widgetJson.optString("encoding", null))
+                .build();
+
+        allWidgets.add(widget);
+
+        JSONArray childWidgetJson = widgetJson.optJSONArray("widgets");
+        if (childWidgetJson != null) {
+            for (int i = 0; i < childWidgetJson.length(); i++) {
+                parseJson(allWidgets, widget, childWidgetJson.getJSONObject(i), iconFormat);
+            }
+        }
     }
 
-    public void setIconColor(String color) {
-        try {
-            this.iconcolor = new Integer(Color.parseColor(fixColorName(color)));
-        } catch(IllegalArgumentException e) {
-            Log.e("OpenHABWidget", "Color was " + color);
-            Log.e("OpenHABWidget", e.getMessage());
-            this.iconcolor = null;
+    private static String determineOH2IconPath(OpenHABItem item, Type type, String icon,
+            String iconFormat, boolean hasMappings) {
+        String itemState = item != null ? item.state() : null;
+        if (itemState != null) {
+            if (item.isOfTypeOrGroupType(OpenHABItem.Type.Color)) {
+                // For items that control a color item fetch the correct icon
+                if (type == Type.Slider || (type == Type.Switch && !hasMappings)) {
+                    try {
+                        itemState = String.valueOf(item.stateAsBrightness());
+                        if (type == Type.Switch) {
+                            if (itemState.equals("0")) {
+                                itemState = "OFF";
+                            } else {
+                                itemState = "ON";
+                            }
+                        }
+                    } catch (Exception e) {
+                        itemState = "OFF";
+                    }
+                }
+            } else if (type == Type.Switch && !hasMappings
+                    && !item.isOfTypeOrGroupType(OpenHABItem.Type.Rollershutter)) {
+                // For switch items without mappings (just ON and OFF) that control a dimmer item
+                // set the state to "OFF" instead of 0 or to "ON" to fetch the correct icon
+                try {
+                    int itemStateNumber = Integer.valueOf(itemState);
+                    if (itemStateNumber == 0) {
+                        itemState = "OFF";
+                    } else {
+                        itemState = "ON";
+                    }
+                } catch (java.lang.NumberFormatException e) {
+                    // Item state is not a number, not sure if that can happen, but good to catch
```

Comment:
Should itemState set to null here?

Reply:
I have no idea, I just copied what was previously present in OpenHAB2Widget.java.

Reply:
Ah, that's my code. The `else if (type == Type.Switch && !hasMappings` branch can be simplified.
````java
if (itemState.equals("0")) {
itemState = "OFF";
} else {
itemState = "ON";
}
````

Reply:
Indeed, or even further: `itemState = itemState.equals("0") ? "OFF" : "ON"; `

Reply:
Done.